covariancematrix.go

```diff
@@ -38,35 +37,42 @@ func CovarianceMatrix(cov *mat64.Dense, x mat64.Matrix, wts []float64) *mat64.De
 	// Subtract the mean of each of the columns.
 	for i := 0; i < c; i++ {
 		v := xt.RawRowView(i)
+		// This will panic with ErrShape if len(wts) != len(v), so
+		// we don't have to check the size later.
 		mean := Mean(v, wts)
 		floats.AddConst(-mean, v)
 	}
 
-	var xc mat64.Dense
-	xc.TCopy(&xt)
-
 	var n float64
 	if wts != nil {
-		if wr := len(wts); wr != r {
-			panic(mat64.ErrShape)
-		}
 
+		// Multiply by the sqrt of the weights, so that multiplication is symmetric.
+		sqrtwts := make([]float64, r)
+		for i, w := range wts {
+			if w < 0 {
+				panic("stat: negative covariance matrix weights")
+			}
+			sqrtwts[i] = math.Sqrt(w)
+		}
 		// Weight the rows.
 		for i := 0; i < c; i++ {
 			v := xt.RawRowView(i)
-			floats.Mul(v, wts)
+			floats.Mul(v, sqrtwts)
 		}
 
 		// Calculate the normalization factor.
 		n = floats.Sum(wts)
-	} else {
-		n = float64(r)
+		cov.MulTrans(&xt, false, &xt, true)
+
+		// Scale by the sample size.
+		cov.Scale(1/(n-1), cov)
+		return cov
 	}
+	n = float64(r)
 
-	cov.Mul(&xt, &xc)
+	cov.MulTrans(&xt, false, &xt, true)
```

Comment:
Since the non-weighted case is simpler and more common, it should probably be 
````
if weights == nil {
 // do simple stuff
}
// Complicated code